### PR TITLE
feat(auth): add configurable scope enforcement with `scope_mode` option

### PR DESCRIPTION
**Pull request by gocamille**
This PR introduces configurable scope enforcement with a new `scope_mode` config option. This replaces the previous "fail-closed" logic where a token required all configured scopes to pass validation.

The new `scope_mode` supports three strategies:
- `require_all` (default): the token must have all configured scopes 
- `require_any`: token must have at least one of the configured scopes
- `disabled`: this skips scope enforcement entirely 

This change addresses requests for more flexible scope matching and explicit support for downstream authorization.

## Changes

- Added `ScopeMode` enum to `auth.rs` with `Disabled`, `RequireAll`, and `RequireAny` variants
- Updated `oauth_validate` middleware to enforce logic based on the selected mode
- Added startup warning when scopes are configured but `scope_mode` is `disabled`

### CODE REVIEW

File: docs/source/auth.mdx [hunk 1]
```diff
@@ -58,6 +58,68 @@ Skipping audience validation reduces security. Only use `allow_any_audience: tru
 
 </Caution>
 
+## Configuring scope enforcement
```
Comment (pragl) [inline docs/source/auth.mdx]:
```suggestion
## Configure scope enforcement
```

File: docs/source/auth.mdx [hunk 2]
```diff
@@ -58,6 +58,68 @@ Skipping audience validation reduces security. Only use `allow_any_audience: tru
...
+
+You can control how the MCP Server validates scopes from OAuth tokens using the `scope_mode` option.
```
Comment (pragl) [inline docs/source/auth.mdx]:
```suggestion
Use the `scope_mode` option to control how the MCP Server validates scopes from OAuth tokens.
```

File: docs/source/auth.mdx [hunk 3]
```diff
@@ -58,6 +58,68 @@ Skipping audience validation reduces security. Only use `allow_any_audience: tru
...
+
+| Mode | Behavior |
+|------|----------|
+| `require_all` | Token must have **all** configured scopes (default) |
```
Comment (pragl) [inline docs/source/auth.mdx]:
```suggestion
| `require_all` | Token must have all configured scopes (default) |
```

File: docs/source/auth.mdx [hunk 4]
```diff
@@ -58,6 +58,68 @@ Skipping audience validation reduces security. Only use `allow_any_audience: tru
...
+    servers:
+      - https://auth.example.com
+    scope_mode: disabled
+```
+
+Skip scope enforcement entirely. The server validates token authenticity (signature and audience) but not scopes.
+
+<Caution>
+
+Use `scope_mode: disabled` only when downstream services (e.g., subgraphs) handle authorization. Without scope enforcement, any valid token grants full access to the MCP Server.
```
Comment (pragl) [inline docs/source/auth.mdx]:
```suggestion
Use `scope_mode: disabled` only when downstream services, such as subgraphs, handle authorization. Without scope enforcement, any valid token grants full access to the MCP Server.
```

**Review by DaleSeo — COMMENTED** [review]
## Review Summary

This PR adds configurable scope enforcement via a new `scope_mode` option, replacing the rigid fail-closed behavior. The implementation is clean, the enum design is idiomatic Rust, and backward compatibility is preserved with `RequireAll` as the default.

### Findings

- **[Should Fix]** Scope validation logic is duplicated between the production middleware and the test helper `is_sufficient`. Extracting into a method on `ScopeMode` would create a single source of truth.
- **[Consider]** Changeset title has a typo: "suport" → "support". This flows into the CHANGELOG.

### Test Coverage Assessment

Test coverage is comprehensive — all three modes are tested, edge cases (empty scopes, no token scopes, partial matches) are covered, and deserialization tests ensure config parsing works. Note that the scope mode tests exercise a reimplemented helper rather than the actual middleware code path; the extraction suggested above would address this.

### Final Recommendation

**Approve with suggestions** — solid, well-scoped change. The duplication finding is the main actionable item.

---
_Reviewed by Claude Code Opus 4.6_

File: .changeset/add_configurable_scope_enforcement_via_scope_mode_to_suport_flexible_matching_strategies_require_any_disabled.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
+default: minor
+---
+
+# Add configurable scope enforcement via `scope_mode` to suport flexible matching strategies (`require_any`, `disabled`)
```
Comment (DaleSeo) [inline .changeset/add_configurable_scope_enforcement_via_scope_mode_to_suport_flexible_matching_strategies_require_any_disabled.md]: **[Consider]** Typo: "suport" → "support". Since changeset titles end up in the CHANGELOG, worth a quick fix.

Comment (DaleSeo) [inline .changeset/add_configurable_scope_enforcement_via_scope_mode_to_suport_flexible_matching_strategies_require_any_disabled.md]:
```suggestion
# Add configurable scope enforcement via `scope_mode` to support flexible matching strategies (`require_any`, `disabled`)
```
